ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/FSRemover.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.util.StringUtils;
+import org.apache.hive.common.util.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+import static org.apache.hadoop.hive.metastore.HMSHandler.getMSForConf;
+import static org.apache.hadoop.hive.metastore.utils.MetaStoreUtils.getDefaultCatalog;
+
+/**
+ * A runnable class which takes in cleaningRequestHandler and cleaning request and deletes the files
+ * according to the cleaning request.
+ */
+public class FSRemover {
+  private static final Logger LOG = LoggerFactory.getLogger(FSRemover.class);
+  private final HiveConf conf;
+  private final ReplChangeManager replChangeManager;
+
+  public FSRemover(HiveConf conf, ReplChangeManager replChangeManager) {
+    this.conf = conf;
+    this.replChangeManager = replChangeManager;
+  }
+
+  public List<Path> clean(CleaningRequest cr) throws MetaException {
+    Ref<List<Path>> removedFiles = Ref.from(new ArrayList<>());
+    try {
+      Callable<List<Path>> cleanUpTask;
+      cleanUpTask = () -> removeFiles(cr);
+
+      if (CompactorUtil.runJobAsSelf(cr.runAs())) {
+        removedFiles.value = cleanUpTask.call();
+      } else {
+        LOG.info("Cleaning as user {} for {}", cr.runAs(), cr.getFullPartitionName());
+        UserGroupInformation ugi = UserGroupInformation.createProxyUser(cr.runAs(),
+                UserGroupInformation.getLoginUser());
+        try {
+          ugi.doAs((PrivilegedExceptionAction<Void>) () -> {
+            removedFiles.value = cleanUpTask.call();
+            return null;
+          });
+        } finally {
+          try {
+            FileSystem.closeAllForUGI(ugi);
+          } catch (IOException exception) {
+            LOG.error("Could not clean up file-system handles for UGI: {} for {}",
+                    ugi, cr.getFullPartitionName(), exception);
+          }
+        }
+      }
+    } catch (Exception ex) {
+      LOG.error("Caught exception when cleaning, unable to complete cleaning of {} due to {}", cr,
+              StringUtils.stringifyException(ex));
+    }
+    return removedFiles.value;
+  }
+
+  /**
+   * @param cr Cleaning request
+   * @return List of deleted files if any files were removed
+   */
+  private List<Path> removeFiles(CleaningRequest cr)
+          throws MetaException, IOException {
+    List<Path> deleted = new ArrayList<>();
+    if (cr.getObsoleteDirs().isEmpty()) {
+      return deleted;
+    }
+    LOG.info("About to remove {} obsolete directories from {}. {}", cr.getObsoleteDirs().size(),
+            cr.getLocation(), CompactorUtil.getDebugInfo(cr.getObsoleteDirs()));
+    boolean needCmRecycle;
+    try {
+      Database db = getMSForConf(conf).getDatabase(getDefaultCatalog(conf), cr.getDbName());
```

Comment:
if the metadata cache contains db as well why do we need to fetch it every time?

Reply:
Added code to fetch database objects from cache if available. Done.